### DISCUSSION TITLE
Overloaded property syntax questions

### DISCUSSION THREAD

**Discussion by hank-bond** [geldata/gel#2721] [Q&A] [answered]
Would love some clarification around what I'm seeing regarding these different scenarios trying to test out this overloaded operator.  I have a very tenuous grasp on EdgeQL as is, so unfortunately there's a good change this just comes down to me mucking it up.

The validity of these schemas is being tested with the `edgedb create-migration --non-interactive` command.

What I expect to work is this schema
```
module test {
    abstract type TestBase {
        required property prop -> str;
    }

    type Test extending TestBase {
        overloaded property prop := .source_prop;
        required property source_prop -> str;
    }
}
```
but it fails validation with the error
```
error: Unexpected ':='
  ┌─ ./dbschema/tese.esdl:7:34
  │
7 │         overloaded property prop := .source_prop;
  │                                  ^^ error
```

Now when removing the `overloaded` keyword I expect it to fail validation with an error (albeit a different one).  It sort of does, but when I validate 
```
module test {
    abstract type TestBase {
        required property prop -> str;
    }

    type Test extending TestBase {
        property prop := .source_prop;
        required property source_prop -> str;
    }
}
```
the command returns 
`edgedb error: ERROR: SchemaError: Property <QualName test::__|prop@test|Test> is already present in the schema <FlatSchema gen:317 at 0x7f9fd8424190>`
which when turning on the interactive mode tells me its an error resulting in a rollback
`Rolling back last operation...`
all good enough I suppose.  Not entirely sure why this couldn't be caught statically, but it's still a very useful error.

Here's a weird thing though, when I run this (what I think is, but might not be) equal schema
```
module test {
    abstract type TestBase {
        required property prop -> str;
    }

    type Test extending TestBase {
        property prop := (select .source_prop);
        required property source_prop -> str;
    }
}
```
the error I get is 
```
error: property 'prop' of object type 'test::Test' must be declared using the `overloaded` keyword because it is defined in the following ancestor(s): test::TestBase
  ┌─ ./dbschema/tese.esdl:7:9
  │
7 │         property prop := (select .source_prop);
  │         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ error
```
Which looks about what I was expecting to receive via the previous schema.  Again, this is all fine and good. 

So my question is unfortunately pretty open ended.  

How should I be using the overloaded operator of this is not the correct way to do it?  

What I want to do is define the `prop` in the abstract class where I can add some Type constraints around a combination of it and other properties, but then define the actual property value its supposed to take on via configuring the child Type.  Kind of like implementing subclasses in OOP languages.

Heres an minimum example of what I'm trying to accomplish

```
module data {
    abstract type Recording {
        constraint exclusive on (
            (._persistent_id, ._version)
        ){
            errmessage := "There cannot be two Recordings with the same version and _persistent_id"
        };

        required property _persistent_id -> str {
            annotation description := "The id that is going to persist from version to version in order to group record lineage";
        };
        required property _version -> int16 {
            annotation description := "Version is incremented anytime a change in the data is registered";
        };       
    }
}
```
and in another module
```
module sources {
    type Hospital extending data::Recording {
        overloaded property _persistent_id := .facility_id ;
        required property facility_id -> str;
    }
}
```

Any help would be greatly appreciated as I make my way through learning EdgeQL.  🙏  The documentation on the `overloaded` keyword is a bit sparse.

**Comment by elprans** [comment, marked as answer]
Hmm.  The handling of the first two scenarios could use some improvement.

Anyway.  The `overloaded` keyword is an SDL-only specifier that you are required to specify in schemas to indicate that you overload a property or a link _knowingly_ and not coincidentally.  We feel it is important to be explicit here for two reasons:  

1) readability: the schema definition informs you that the link/property is defined in a supertype and is being modified in a subtype in some way;
2) migration correctness: if the upstream schema you are inheriting from drops the property or link in question, your overload would become invalid, and you'd have to make an explicit decision to drop it or make it non-overloaded (instead of implicitly continuing to define a potentially bogus non-inherited property).

Now, to address your specific use case.  There is an unfortunate limitation currently: you cannot overload regular properties with computed properties, or vice-versa, or overload any computed property with another computed property.  This limitation is temporary and is only there because, at the moment, our query compiler is unable to handle computable property dispatch properly when you query the parent (abstract) type.  For example, if you do `SELECT Recording { _persistent_id }`, we need to explicitly compute and `UNION` all computed expressions, perhaps mixing them with non-computed variants.  We'll be addressing this limitation soon, but for now you'd have to either copy the `annotation` definition or avoid making `_persistent_id` a computed property.

**Reply by hank-bond** [reply]
So I guess the only solution for this is to keep `_persistent_id` only in the abstract class as a static property, and specify the value on the data/insertion side instead of computing it with a reference.  This seems less optimal because the logic for that is now in my processing script and not in the schema.  My favorite part of EdgeDB so far has been how much logic can be coded into the database/schema itself (expression constraints are amazing).  On the plus side this seems like a feature I can migrate towards when available.

Thank you for the clarification!

**Reply by hank-bond** [reply]
Oh by the way, the necessity of using the overloaded keyword is an intuitive design choice that I agree is a good idea.  There was no confusion on my end as to "why" that needed to happen, just the "how".
